### DISCUSSION TITLE
Multiplatform Lets-Plot API

### DISCUSSION THREAD

**Discussion by alshan** [JetBrains/lets-plot-kotlin#73] [Show and tell]
cc: @sureshg, @Waschnick, @LouisCAD, @holgerbrandl, @icuxika, @bjonnh, @veyndan, @descent2oblivion

I've lately been experimenting with lets-plot and Kotlin JS frontend.

I converted "plot API" to multi-platform project (currently in the branch [api-mp](https://github.com/JetBrains/lets-plot-kotlin/tree/api-mp)) and published the following MP artifacts to sonatype snapshots repository:
- lets-plot-kotlin 
Use as dependency in Kotlin MP project ('common' source-set).
- lets-plot-kotlin-js 
Use in Kotlin JS project.
- lets-plot-kotlin-jvm
Use in Kotlin JVM project. 
This artifact will replace the current "lets-plot-kotlin-api" artifact.
- lets-plot-kotlin-metadata

I've also created a Github repository containing 3 small demos configured to use these artifacts: JS app, JVM/Batik and JVM/JavaFX apps: https://github.com/alshan/lets-plot-mini-apps

Everything worked fine to me but these apps are very simple and may be I'm still missing something. Could you try Lets-Plot MP artifacts in your applications?

**Comment by descent2oblivion** [comment]
Hi there!

I updated my app to use:
   ` api "org.jetbrains.lets-plot:lets-plot-kotlin-api:2.0.1"`
instead of:
`implementation "org.jetbrains.lets-plot:lets-plot-kotlin-api:0.0.26-SNAPSHOT"`

The only issue I got is the following:
![imagen](https://user-images.githubusercontent.com/58195806/119741865-6ca48000-be3b-11eb-90b0-259669b92ff4.png)

If I remember the version of the lets-plot API downloads the script and executes locally. Is there documentation available on how can I access the script?

Всего доброго!! )))

**Reply by alshan** [reply]
Hi, you now have to specify the script version via the `scriptUrl` param:
```
scriptUrl = PlotHtmlHelper.scriptUrl("2.0.1")
```
Удачи )

**Comment by descent2oblivion** [comment]
Спасибо большое!
I did as you said, but the plot doesn't show. With help of the debbuger I found out the following issue:

On the first square it seems there is a problem with the script.

Then on the second square is highlighted the URL generated by the code:
`scriptUrl = PlotHtmlHelper.scriptUrl("2.0.1")`

Finally the third square is the exact line where the script fails.

![lp_notDefined](https://user-images.githubusercontent.com/58195806/119767922-a346be80-be6c-11eb-9a1a-0849cabe3881.png)

As an alternative I exctracted the URL from the previous version:
Version: `implementation "org.jetbrains.lets-plot:lets-plot-kotlin-api:0.0.26-SNAPSHOT"`
URL: `"https://dl.bintray.com/jetbrains/lets-plot/lets-plot-latest.min.js"`

And implemented it like this:
![old_CDN](https://user-images.githubusercontent.com/58195806/119768503-97a7c780-be6d-11eb-9b97-291ecc5d09fc.png)

This resulted in the plots rendering again.

So my guess is there is something going on with the 
`scriptUrl = PlotHtmlHelper.scriptUrl("2.0.1")`

**Reply by descent2oblivion** [reply]
I've updated my repo of the app to help in the error replication )))

https://github.com/descent2oblivion/plot-app

**Reply by alshan** [reply]
Oh, sorry, "2.0.1" is the Kotlin API version. JS should be 2.0.2:
```
scriptUrl = PlotHtmlHelper.scriptUrl("2.0.2")
```
Lets-Plot is distributed via CDNJS since 2.0.2: https://cdnjs.com/libraries/lets-plot/2.0.2

**Reply by descent2oblivion** [reply]
Than you so much!
Everything now works perfectly )))
I have updated the repo too!!

Спасибо большое и всего доброго!! 😄

**Reply by alshan** [reply]
Wait, wait) Could you try the experimental "multi-platform" artifact with your app? 

- remove the "org.jetbrains.lets-plot:lets-plot-kotlin-api:2.0.1" dependency (it's gonna be no more)
- add the new dependency: "org.jetbrains.lets-plot:lets-plot-kotlin-jvm:2.0.2-SNAPSHOT"
- add "SNAPSHOT" maven repository:
```
        maven {
            url "https://oss.sonatype.org/content/repositories/snapshots"
        }
```

I've put out [lets-plot-mini-apps](https://github.com/alshan/lets-plot-mini-apps) repo as an example.

**Reply by descent2oblivion** [reply]
Yes, of course, I've already updated it and works perfectly :)))
I am sharing some pics of my app

![1](https://user-images.githubusercontent.com/58195806/119935639-ac9f5c00-bf3c-11eb-9b40-36e16eb48ecd.jpeg)
![2](https://user-images.githubusercontent.com/58195806/119935638-ac9f5c00-bf3c-11eb-8681-b0a0a8e421dc.jpeg)
![3](https://user-images.githubusercontent.com/58195806/119935637-ac06c580-bf3c-11eb-807c-c1d9e816c24c.jpeg)

I've updated the repo as well 
https://github.com/descent2oblivion/plot-app

Thank you for the support!

**Reply by tigerjenny** [reply]
Excuse me, are you developing with android studio?

I download your project and use android studio to run, the project can work.
I create a new one and use higher version, but it can not show the html result on webView.😰

**Reply by descent2oblivion** [reply]
Hello there, I apologize the late reply.
I saw you already solve the issue, I'll leave some tips in case someone needs help in the future:

Lets first state that the api required the internet permission in the **AndroidManifest.xml** so we need to add the following line

`<uses-permission android:name="android.permission.INTERNET"/>`

TIP 1
Try display the html value on a field so you can copy it to your clip board, with this HTML you can make a sample .html file which you can run on your browser.

The expected result is that your plot displays correcly planned. If the plot doesn't come up maybe you can use the browser console to display the errors, the only issue I faced was when I mistyped the helper version, which is a cdn js file containing the required functions to assemble the plot.

The solution is to check you are using a working version of the helper for my sample I used "2.0.2" but you can use the one that best fits your requirements.

TIP 2

Once you verified the results of TIP 1 we can conclude that in fact the app is working the plotting process, so we need to test the display (WebView) itself.
What I recommend is try to load a sample site that you are 100% sure it loads quick and safe, let's say github.com.

The expected result is the site to load as normal.

If the site doesn't load or loads partially then its recommended for you to check the WebView properties, by default settings it is likely the WebView doesn't load. You need to set the javaScriptEnabled property to True (Shown bellow) 

`wevi.getSettings().javaScriptEnabled = true`  (wevi is the name you gave to your WebView)


These are the most common issues you can face when working with the api. I hope these tips are helpful

Once again I apologize for the delay. kudos :)

**Reply by tigerjenny** [reply]
Thank you so much! 😁

**Reply by alshan** [reply]
@descent2oblivion @tigerjenny , thank you for valuable input!

**Reply by alshan** [reply]
https://github.com/JetBrains/lets-plot-kotlin/commit/d0505e2206dfde27bc0c530bad5b5b7edd717fb2

**Comment by alshan** [comment]
Guys, Lets-Plot Kotlin went "multi-platform" in [v3.0.0](https://github.com/JetBrains/lets-plot-kotlin#what-is-new-in-300) !

**Reply by maqui7295** [reply]
Hi @alshan. Thank you for the awesome Lets-plot.

I've been working on a project (yet to be hosted) to provide a free web UI to lets plot API to help Researchers who can't code. 
Though I'm currently using the Python API, I was wondering if I could embed Lets-Plot directly in a **native (either C or Rust) desktop project** without going the Python route or JVM.

Maybe via kotlin-native. Do you mind giving me some pointers?

Thank you.

**Reply by alshan** [reply]
Hi Mark, happy to hear you like Lets-Plot!

The vast majority of LP Kotlin API is in 'common' target (in terms of Kotlin-multiplatform). Thus you can use it in a Kotlin-native project.

For plot rendering however, we support only JVM and browser. So a Kotlin-native app is only possible if there is some kind of embedded browser available, or at the very minimum an SVG viewer.

**Reply by maqui7295** [reply]
That means I can create SVG via Kotlin-native and use a Rust framework like dioxus (which uses a webview for desktop). Nice!

Thanks Alshan.

**Reply by alshan** [reply]
Sure, showing a static SVG shouldn't be a problem I guess. For an interactive plot however you'll need to craft a little web-page that includes LP js library.

**Reply by maqui7295** [reply]
Yeah, a static SVG is very okay for many Academic researchers. 
I just tried LP js. It's great. One can even share a plot spec! Thanks for the tip.

**Reply by maqui7295** [reply]
Also, I'd be happy to contribute to documentation if you need people to.

**Reply by alshan** [reply]
Certainly, such contributions are very welcome! It could be anything from fixing a typo to adding cookbook-style notebooks. You can contribute through issues, discussions, or pull requests - whatever works best for you.

**Reply by maqui7295** [reply]
Okay.
